New Sentry: Add JWKS validator

ItalyPaleAle opened the pull request:
This is a PR against dapr/dapr#6171 which adds support for an alternative token validator called JWKS.

These are the main changes:

1. Renamed the "selfhosted" validator to "insecure", since it doesn't do any validation at all.
1. Added a new `mtls.tokenValidators` option to the "daprsentry" Configuration resource to configure *additional* validators. This is an array that contains that contains:
   - `mtls.tokenValidators[*].name`: name of the validator.
   - `mtls.tokenValidators[*].options`: optional dictionary of key-values that are options that can be passed to validator.
1. In Kubernetes mode, the built-in "kubernetes" validator is always enabled.  In self-hosted mode, the built-in "insecure" validator is enabled only if there's no additional validator configured.
1. Added the `jwks` validator that validates . 
   - This accepts 1 option: `source` is the location of a JWKS that contains the **public** keys to validate the keys. This can be one of:
     - A path to a file on disk - this is watched for changes with fsnotify
     - A HTTP(S) URL - this is periodically refreshed if needed
     - The actual JWKS in the string value, optionally base64-encoded
   - This validator requires JWT tokens that contain the 2 claims that SPIFFE's JWT-SIV use too:
     - `sub` must be the SPIFFE ID of the app, for example `spiffe://localhost/ns/default/myapp`
     - `aud` must contain the SPIFFE ID of Sentry, for example `spiffe://localhost/ns/dapr-system/dapr-sentry`
     - The validator will validate time validity bounds if present, but it will not enforce their presence (so tokens can live as long as the issuer wants them to)
1. `daprd` can load a token to send to Sentry that is signed with an arbitrary key. These are defined using:
   - A file pointed by the env var `DAPR_SENTRY_TOKEN_FILE`
   - (Note that the env var must start with `DAPR_` for security reasons, since those are not accessible in the localenv secret store)
1. Token requests that want to use a non-default validator must pass the "token_validator" option in the gRPC request.

## To test

To test this, go into the folder `cmd/sentry`.

Create the file `config.yaml` (includes the JWKS inlined):

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: daprsystem
  namespace: default
spec:
  mtls:
    enabled: true
    tokenValidators:
      - name: "jwks"
        options:
          source: |
            {"keys":[{"kid":"mykey","alg":"RS256","kty":"RSA","use":"sig","e":"AQAB","n":"3I2mdIK4mRRu-ywMrYjUZzBxt0NlAVLrMhGlaJsby7PWTMiLpZVip4SBD9GwnCU0TGFD7k2-7tfs0y9U6WV7MwgCjc9m_DUUGbE-kKjEU7JYkLzYlndys-6xuhD4Jf1hu9AZVdfXftpWSy_NNg6fVwTH4nckOAbOSL1hXToOYWQcDDW95Rhw3U4z04PqssEpRKn5KGBuTahNNNiZcWns99pChpLTxgdm93LjMBI1KCGBpOaz7fcQJ9V3c6rSwMKyY3IPm1LwS6PIs7xb2ZJ0Eb8A6MtCkGhgNsodpkxhqKbqtxI-KqTuZy9g4jb8WKjJq9lB9q-HPHoQqIEDom6P8w"}]}
          minRefreshInterval: 2m
          requestTimeout: 600s
```

The corresponding private key (if you need to generate JWTs) is:

```text
-----BEGIN RSA PRIVATE KEY-----
MIIEpQIBAAKCAQEA3I2mdIK4mRRu+ywMrYjUZzBxt0NlAVLrMhGlaJsby7PWTMiL
pZVip4SBD9GwnCU0TGFD7k2+7tfs0y9U6WV7MwgCjc9m/DUUGbE+kKjEU7JYkLzY
lndys+6xuhD4Jf1hu9AZVdfXftpWSy/NNg6fVwTH4nckOAbOSL1hXToOYWQcDDW9
5Rhw3U4z04PqssEpRKn5KGBuTahNNNiZcWns99pChpLTxgdm93LjMBI1KCGBpOaz
7fcQJ9V3c6rSwMKyY3IPm1LwS6PIs7xb2ZJ0Eb8A6MtCkGhgNsodpkxhqKbqtxI+
KqTuZy9g4jb8WKjJq9lB9q+HPHoQqIEDom6P8wIDAQABAoIBAQCwSSxWDjI8cR+v
9WqAvlB/1jHeeKjgIQXK9F3QucHC1RCiON2d0USUfw2ltkZC2F3v+wKQFGZbhA1Q
U5G8KBgMFwfrj25Sgp55VrmaaLrb5z5xGfDIe9sbxNsbzPp7awbWVqQZgDdHtn6p
BCEzZwanSDP4aKBzg8UYUP7kxEXAQwc/Z/L7U8SnZ0oGJPeyXbKjgiMw9FYxaHV7
/bGe8VdhIYLHX7WrIP9WkIKkvv6S5NYJkU7HwKr8SlmJfui8NUERvQmWtPeytcAC
mQQTjZhgWeXdN6OWBk7LTSC/up77pYHXuUEz/b94V3mWJ0nNVLiuInFloE0l5bSy
D9K6xgARAoGBAPSzprPH3dBVcIdgsknMQVs+T4O4FbZQpuqg9mSg2S/OBQw/wVcs
uS0JvZVeJE3BN08lh+nR3hpieIN1habmOMbrLlPK3VVDr/naXC6FuICZ0OLGqinY
t7lOU5s6gqr8zV0qBAr8uQD1yvi4dNXOX4K5nkc2FBCTdoYMcWi+hyQZAoGBAOa8
kSiac7GQ9gdcW521HkS04wUajiKhbFbhdJTIu0jGduCLCiH2vQngQnxiBZSMBnvo
yG6My9payLMN0N42R3MCFavatbPb6GXluGe3MpQgbPS6udffW+fuvkLngyuBZLIO
bkLA2NiZXmppudy+AjCI5gUWlRFDFu40X8TUCXXrAoGBAL2/bR3tdugsyJSBe97T
3z4kpD3sihGUmspHxKglPiTeqXVlHsQ9bt4otnTTymzszbm2zZyXW5W0E+VERT41
OYOuLH9XSKfsU9H37BE7KeW24I5pJNxMK7MqtJQjG94BKDH8+aOXctH2xsx1JEwC
undA/iYUHhz4bCjoDyz8yQhJAoGBAKZE1+2Ebz5TkZb4TOS2cVtdZgJ1glseK63n
xVS52Yu1yFBCoUicPtjxC7gKAq9qv/lmY5iWSQfOR7YPjKW0Q4Ryyh8O+cTkAPH8
ugezEfV/8EPikq0mvVnCxsP7hKBDG8bbc0UTJq/exRS1n0OcgINtAyRDg6PuzoVU
pHZ3vWGFAoGAc5Um3YYkhh2QScQBy5+kumH40LhFFy2ETznWEp0tS2NwmTfTm/Nl
Sg+Ct2nOw93cIhwDjWyoilkIapuuX2obY+sUc3kj2ugU+hONfuBStsF020IPP1sk
A9okIZVbz8ycqcjaBiNc4+TeiXED1K7bV9Kg+A9lxDxfGRybJ1/ECWA=
-----END RSA PRIVATE KEY-----
```

Then you can invoke the gRPC `SignCertificate` method. Example request payload (as JSON):

```json
{
    "id": "spiffe://localhost/ns/default/myapp",
    "namespace": "default",
    "token_validator": "jwks",
    "certificate_signing_request": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURSBSRVFVRVNULS0tLS0KTUlIOU1JR2xBZ0VBTUJneEZqQVVCZ05WQkFNVERXaGxiR3h2WW14MVpXUmhjSEl3V1RBVEJnY3Foa2pPUFFJQgpCZ2dxaGtqT1BRTUJCd05DQUFTNnRQWC9VOHd2YlV1b2FsbHZrWHdFRTVVRm1mK3E5Ujl5U2FhcGQ2TU9tL0NCClNhMzZIY2hYbjJrZlVKOENZMm5VbXNuUDIrSzZibmJoMWZxQWZhZU1vQ3N3S1FZSktvWklodmNOQVFrT01Sd3cKR2pBWUJnTlZIUkVFRVRBUGdnMW9aV3hzYjJKc2RXVmtZWEJ5TUFvR0NDcUdTTTQ5QkFNQ0EwY0FNRVFDSUdQZAo5bUZ5RWhVUnFKZ3VRNW9aNVdKcTFJb2dMdDBWb1ZhMHBuV21OVTVIQWlBRTJydnJaZFlwWURITUpQYzQrVnV0CkpyMGN4bHdVM29FOG5UT0pDL3VweGc9PQotLS0tLUVORCBDRVJUSUZJQ0FURSBSRVFVRVNULS0tLS0=",
    "token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Im15a2V5In0.eyJpYXQiOjE1MTYyMzkwMjIsImF1ZCI6WyJzcGlmZmU6Ly9sb2NhbGhvc3QvbnMvZGVmYXVsdC9kYXByLXNlbnRyeSJdLCJzdWIiOiJzcGlmZmU6Ly9sb2NhbGhvc3QvbnMvZGVmYXVsdC9teWFwcCJ9.2NNsBk6ysWa7RDjitOt_QLPxAUgUtniVnrGVuv5-vp-doq0y8Ocwiufy983PLTgdA8pDfSROY-387QVtIs7Xc49WzZ68f1F5-SfWqhJh_tmB6pgw5mU9TMRPh8Cmrx6Vy3oKDB01gKkbNPfDj_D8pKha3aMoFrPOrM26lhAhkw18R6YOrMWJyiNFc2ONtyQ08seREtjlyJs5fpKC2S6MvwYQQq2HgSr--2cUbSN1LUmC38eL2YdCz1id39UIWt74CumzWEaRN0qDs2-tnztAgKRIixtd7oJqGSF8w8CXRbaJre2jk6gb9MuFKTyE9RuVOUxWpFDQDuZZ9WyfutmO6Q"
}
```

The included JWT's data is:

![image](https://github.com/JoshVanL/dapr/assets/43508/9a026774-111e-4c9b-9557-1c831c3d5012)

> I use https://jwt.io/ to generate the signed JWTs

Start sentry with:

```sh
mkdir -p /tmp/sentry
go run . -config config.yaml -issuer-credentials /tmp/sentry
```

ItalyPaleAle commented on pkg/runtime/security/auth.go:
```diff
@@ -81,7 +80,7 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
 		return nil, err
 	}
-	certPem := pem.EncodeToMemory(&pem.Block{Type: certType, Bytes: csrb})
+	certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrb})
```
They said "CERTIFICATE" before but they are CSRs

ItalyPaleAle commented on pkg/security/fake/fake.go:
```diff
@@ -1,3 +1,6 @@
+//go:build unit
```
I added these so the files don't accidentally get built into the final binary. `unit` is the tag we use for all code that is only in unit tests

ItalyPaleAle commented on pkg/sentry/options.go:
```diff
@@ -0,0 +1,128 @@
+/*
```
This file was copied almost as-is from components-contrib. We may need to move it to dapr/kit eventually

JoshVanL replied:
I think moving to `/kit` now makes sense.

ItalyPaleAle commented on pkg/sentry/sentry.go:
```diff
@@ -104,45 +108,43 @@ func (s *sentry) Start(ctx context.Context) error {
 	}
 
-	validatorErr := make(chan error, 1)
+	errCh := make(chan error, 3)
```
Without this change on how errors are returned, I couldn't get errors from `val.Start()` to make the app exit. It would just hang until SIGTERM.

ItalyPaleAle commented on pkg/sentry/server/validator/insecure/insecure.go:
```diff
@@ -23,17 +23,19 @@ import (
-type selfhosted struct{}
+// insecure implements the validator.Interface. It doesn't perform any authentication on requests.
+// It is meant to be used in self-hosted scenarios where Dapr is running on a trusted environment.
+type insecure struct{}
```
If you have a better name pls suggest!

daixiang0 replied:
LocalValidator?

ItalyPaleAle replied:
The thing is:

- it doesn’t do any validation at all. It accepts anything
- JWKS can be local too (JWKS works on both self-hosted and kubernetes)

JoshVanL replied:
I think `insecure` is the best name (and definitely correct!)

ItalyPaleAle commented:
Completed the changes to add the additional validators as optional. Now the caller must specify the name in "token_validator"  if it wants to use a non-default token validator.

The token validator works. However, in K8s mode Sentry isn't issuing the certificates to apps because the request contains a `req.Id` that is the service account name. This is changing, and for JWT-based auth backwards compatibility doesn't matter anyways.

JoshVanL reviewed (changes requested):
I have not completed the code review @ItalyPaleAle , but have some comments on the design so you can get feedback earlier :slightly_smiling_face: 

- `Renamed the "selfhosted" validator to "insecure", since it doesn't do any validation at all.` :+1: 
- `location`: `source` seems like a more appropriate name here. We can then use a protocol scheme to select _how_ to source, i.e. `file://`, `https://` etc. We should never accept in line- this puts secrets in clear text in _configuration_(!).
- `sub must be the Dapr app ID, for example myapp`: This claim seems like it should be a SPIFFE ID, and at least _definitely_ contain the namespace.
- `The validator will validate time validity bounds if present, but it will not enforce their presence (so tokens can live as long as the issuer wants them to)`: this seems bad. Why would we not require incoming tokens to be time bound, else we are just encouraging bad security.
-  `A value in the env var DAPR_SENTRY_TOKEN`: We should only accept file references I think- this again going to encourage putting secrets in configuration which is already a problem today we are trying to get  away from. This also means the tokens cannot be changed without completely restarting the daprd process.
- `A file pointed by the env var DAPR_SENTRY_TOKEN_FILE` Can we at least have a CLI flag counter part? This think this is a larger problem with dapr flags/env vars in general though, and I would like to have some standardization project wide around making sure things can be flags and env vars, or just flags. I would like to spend some time on making component options consistently defined and parsed across the project.

ItalyPaleAle commented:
Thanks for the review @JoshVanL 

>     * `location`: `source` seems like a more appropriate name here. We can then use a protocol scheme to select _how_ to source, i.e. `file://`, `https://` etc. We should never accept in line- this puts secrets in clear text in _configuration_(!).

I agree `source` sounds better. I'll make the change. I can also add `file://` at the beginning, that's not a bad idea either.

As for in-line, the value is not a secret (the JWKS only contains public keys), so it should be fine IMHO.

>     * `sub must be the Dapr app ID, for example myapp`: This claim seems like it should be a SPIFFE ID, and at least _definitely_ contain the namespace.

How it's implemented so far, it matches that `req.Id` is what is included in `sub`. For Dapr in K8s right now, that actually means the service account name - which is really bad. But then, whatever you (meaning, you @JoshVanL :) ) choose the include in the req.Id is what the validator will require.

>     * `The validator will validate time validity bounds if present, but it will not enforce their presence (so tokens can live as long as the issuer wants them to)`: this seems bad. Why would we not require incoming tokens to be time bound, else we are just encouraging bad security.

I agree in principle, but I think this will be up to the IdP that creates the JWTs to decide. If the token is passed as env var, for example, it cannot be rotated without killing the pod, which is not an acceptable thing in many cases.

>     * `A value in the env var DAPR_SENTRY_TOKEN`: We should only accept file references I think- this again going to encourage putting secrets in configuration which is already a problem today we are trying to get  away from. This also means the tokens cannot be changed without completely restarting the daprd process.

Yup, that's what I was writing above :) I am open to this idea, but I need to validate with our users if this is fine with them.

>     * `A file pointed by the env var DAPR_SENTRY_TOKEN_FILE` Can we at least have a CLI flag counter part? This think this is a larger problem with dapr flags/env vars in general though, and I would like to have some standardization project wide around making sure things can be flags and env vars, or just flags. I would like to spend some time on making component options consistently defined and parsed across the project.

The advantage of an env var is that in K8s when using the injector, it can be set by adding the annotation `dapr.io/env: "DAPR_SENTRY_TOKEN_FILE=..."`. Using CLI flags would require creating a separate annotation too.

JoshVanL commented on pkg/sentry/server/validator/jwks/jwks.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	// Validate the authorization token
+	_, err = jwt.Parse([]byte(req.Token),
```
We must also be checking the namespace here.

ItalyPaleAle replied:
Makes sense

JoshVanL replied:
Where are we checking the namespace? To me it looks like we are only checking the App ID, which is not good enough and is a massive security vulnerability.

ItalyPaleAle replied:
Done

JoshVanL commented on .golangci.yml:
```diff
@@ -136,6 +136,8 @@ linters-settings:
       - "github.com/golang-jwt/jwt": "must use github.com/lestrrat-go/jwx/v2"
       - "github.com/golang-jwt/jwt/v4": "must use github.com/lestrrat-go/jwx/v2"
       - "github.com/ghodss/yaml": "must use sigs.k8s.io/yaml"
+      # Commonly auto-completed by gopls
```
:+1:

JoshVanL commented on dapr/proto/sentry/v1/sentry.proto:
```diff
@@ -34,6 +34,8 @@ message SignCertificateRequest {
   // A PEM-encoded x509 CSR.
   bytes certificate_signing_request = 5;
+  // Name of the validator to use, if not the default for the environemtn. 
+  string token_validator = 6;
```
Do we want this to be an enum? I think it is useful for us to be putting the supported token validators as part of the Dapr versioned API which is strongly typed and validated. We should put enum value constant `UNKNOWN = 0` so sentry can catch validators it doesn't know yet.

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/acl/acl.go:
```diff
@@ -32,6 +32,10 @@ import (
 	commonv1pb "github.com/dapr/dapr/pkg/proto/common/v1"
 )
 
+const (
+	SpiffeIDPrefix = "spiffe://"
+)
+
```
FYI, this is all going to be chucked out, in favor of `github.com/spiffe/go-sdk/v2`

JoshVanL commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -113,11 +113,23 @@ type HandlerSpec struct {
+// ValidatorSpec contains additional token validators to use.
+type ValidatorSpec struct {
+	// Name of the validator
+	Name string `json:"name"`
```
Enum the string, with constants pointing to each validator name.

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/config/configuration.go:
```diff
@@ -226,19 +226,35 @@ type NameResolutionSpec struct {
+// ValidatorSpec contains additional token validators to use.
+type ValidatorSpec struct {
+	// Name of the validator
+	Name string `json:"name"`
```
Same, re-enum

ItalyPaleAle replied:
This is not used to generate any CRD, so it can't have enums here

JoshVanL commented on pkg/injector/sidecar/container.go:
```diff
@@ -317,7 +317,7 @@ func GetSidecarContainer(cfg ContainerConfig) (*corev1.Container, error) {
 		},
 		corev1.EnvVar{
-			Name:  "SENTRY_LOCAL_IDENTITY",
+			Name:  authConsts.SentryLocalIdentityEnvVar,
```
`SENTRY_LOCAL_IDENTITY` is a bad name for this environment variable, and doesn't event start with `DAPR`. I also don't understand the use case for this env var anyway- the dapr identity is it's namespace and app ID (along with trust domain). I think we should remove this entirely.

ItalyPaleAle replied:
I agree it’s bad but please don’t remove it until you can be sure that won’t break some scenarios. Backwards compatibility first of all!

JoshVanL replied:
Indeed, we'll need it for a release but it won't be used from v1.12

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -17,15 +17,18 @@ import (
 	legacyKubeTknPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
 	sentryMaxRetries  = 100
+
+	jwksValidator       = "jwks"
+	kubernetesValidator = "kubernetes"
```
Enum

ItalyPaleAle replied:
Using the enum from protobuf now

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -109,11 +112,13 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
 	resp, err := c.SignCertificate(context.Background(),
 		&sentryv1pb.SignCertificateRequest{
 			CertificateSigningRequest: certPem,
 			Id:                        getSentryIdentifier(id),
```
This should always be the app ID, but this package is going to be ripped out anyway.

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +158,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+// Returns the token for authenticating with Sentry.
+func getToken() (token string, validator string) {
+	// Check if we have a token in the DAPR_SENTRY_TOKEN env var (for the JWKS validator)
+	if v := os.Getenv(consts.SentryTokenEnvVar); v != "" {
+		log.Debug("Loaded token from DAPR_SENTRY_TOKEN environment variable")
+		return v, jwksValidator
+	}
```
I strongly think that we should not have the option for users to define static secrets in env vars, since env vars are often in the "config" rather than "secret" bucket, so prone to leakage, particularly in Kubernetes environments. This also means that dapr must restart in order to pick up a new token.

I think we should remove this option completely.

ItalyPaleAle replied:
I think there’s use for env vars, including self-hosted scenarios and development. Maybe just discourage them?

JoshVanL replied:
There isn't really a scenario where a user can set an env var but not write a file and reference it. Given the env vars are static, I def think we should be removing support here.

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +158,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+			log.Debugf("Loaded token from path '%s' environment variable", path)
+		}
+		return string(b), jwksValidator
+	}
```
We should change this to:

1. Check if the env var is defined at all (`os.LookupEnv`), which if it isn't this code block is skipped. If the env var is not defined, then any missing token should result in an error in this block.
2. Read token from file, returning any error, including if the env var value is empty.
3. Return read token

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -163,12 +188,16 @@ func getToken() string {
 func getSentryIdentifier(appID string) string {
 	// return injected identity, default id if not present
-	localID := os.Getenv("SENTRY_LOCAL_IDENTITY")
+	localID := os.Getenv(consts.SentryLocalIdentityEnvVar)
```
All relevant changes here also need to be made in `pkg/security/x509source.go`. `pkg/security` will eventually completely supersede this package in a later PR, but until then we need to duplicate functionality during that transition. I pain I know, but we need to cut up the mega diff somewhere :grimacing:

ItalyPaleAle replied:
Looks like `pkg/security/x509source.go` is up-to-date with these changes

JoshVanL commented on pkg/runtime/security/consts/consts.go:
```diff
@@ -15,9 +15,16 @@ package consts
+	// SentryLocalIdentityEnvVar is the environment variable for the local identity sent to Sentry.
+	SentryLocalIdentityEnvVar = "SENTRY_LOCAL_IDENTITY"
+	// SentryTokenEnvVar is the environment variable for the Sentry token.
+	SentryTokenEnvVar = "DAPR_SENTRY_TOKEN"
```
Remove

JoshVanL commented on pkg/runtime/security/consts/consts.go:
```diff
@@ -15,9 +15,16 @@ package consts
+	APITokenEnvVar = "DAPR_API_TOKEN"
+	// AppAPITokenEnvVar is the environment variable for the app API token.
 	AppAPITokenEnvVar = "APP_API_TOKEN"
+	// SentryLocalIdentityEnvVar is the environment variable for the local identity sent to Sentry.
+	SentryLocalIdentityEnvVar = "SENTRY_LOCAL_IDENTITY"
```
Any good reason not to remove? See comment above?

ItalyPaleAle replied:
I am not sure. I don't know what this could break

JoshVanL commented on pkg/runtime/security/consts/consts.go:
```diff
@@ -15,9 +15,16 @@ package consts
 const (
-	// APITokenEnvVar is the environment variable for the api token.
-	APITokenEnvVar    = "DAPR_API_TOKEN"
+	// APITokenEnvVar is the environment variable for the API token.
```
```suggestion
	// APITokenEnvVar DAPR_API_TOKEN is the environment variable for the API token. If set, Dapr will enforce that all incoming API requests has this value set on the `dapr-api-token` HTTP Header. 
```

JoshVanL commented on pkg/sentry/config/config.go:
```diff
@@ -49,6 +50,18 @@ const (
+type ValidatorName string
+
+const (
+	// Kubernetes validator (default on Kubernetes).
+	ValidatorKubernetes ValidatorName = "kubernetes"
+	// Insecure validator (default on self-hosted).
+	ValidatorInsecure ValidatorName = "insecure"
+	// JWKS validator
+	ValidatorJWKS ValidatorName = "jwks"
```
Nit: 
```suggestion
	// Insecure validator (default on self-hosted).
	ValidatorInsecure ValidatorName = "insecure"
	
	// Kubernetes validator (default on Kubernetes).
	ValidatorKubernetes ValidatorName = "kubernetes"
	
	// JWKS validator
	ValidatorJWKS ValidatorName = "jwks"
```

ItalyPaleAle replied:
These constants have been removed since the enum is in the protos now

JoshVanL commented on pkg/sentry/config/config.go:
```diff
@@ -172,5 +179,35 @@ func parseConfiguration(conf Config, daprConfig *daprGlobalConfig.Configuration)
 	conf.Features = daprConfig.Spec.Features
 
+	// Get token validators
+	// In Kubernetes mode, we always allow the built-in "kubernetes" validator
```
We shouldn't allow the `insecure` validator when in Kubernetes mode imo.

ItalyPaleAle replied:
Yeah it shouldn’t be allowed. I don’t believe the current code allows that? Unless I’m mistaken.

JoshVanL replied:
What happens if we are running in Kubernetes, and the `insecure` validator is configured? We should error in this case.

ItalyPaleAle replied:
In K8s mode, the `insecure` validator is not allowed. See below: https://github.com/JoshVanL/dapr/pull/1/files#diff-c818cbb5d734e017aeccd811d955d15d5ded29b0092fc9026e6ab79726577324R192-R193

JoshVanL commented on pkg/sentry/config/config.go:
```diff
@@ -172,5 +179,35 @@ func parseConfiguration(conf Config, daprConfig *daprGlobalConfig.Configuration)
+		conf.DefaultValidator = ValidatorInsecure
+		conf.Validators[ValidatorInsecure] = map[string]string{}
+	}
+
```
If insecure is enabled, we should log a big error warning that this is the case, and that the user should not be using this in production environments.

ItalyPaleAle replied:
Except that’s how Dapr runs when not in K8s today :D We have lots of users who use Docker Compose for example, so they either not use mTLS, or they use mTLS with the “insecure” validator. Neither is great.

JoshVanL commented on pkg/sentry/sentry.go:
```diff
@@ -104,60 +108,96 @@ func (s *sentry) Start(ctx context.Context) error {
 	sec, err := provider.Handler(ctx)
 	if err != nil {
-		return errors.Join(err, <-providerErr, <-validatorErr)
+		return err
```
Here we miss any errors sent to `errCh`. We also will not be waiting for all go routines to return.

JoshVanL replied:
This block should be moved to line 115

ItalyPaleAle replied:
Before there was a situation where errors returned by the init of the validator were not reported. Things just hanged.

JoshVanL commented on pkg/sentry/sentry.go:
```diff
@@ -104,60 +108,96 @@ func (s *sentry) Start(ctx context.Context) error {
-	})
-	return errors.Join(err, <-providerErr, <-validatorErr)
+	// On the first error, cancel the context and return
+	return <-errCh
```
We will only be returning the first error here, regardless if there were 2 or more.

Since we also need to propagate a closed context if _any_ routine returns error, we should use the `concurrency.RunnerManager` here.

ItalyPaleAle replied:
I'm not sure I understand what you mean with `concurrency.RunnerManager`. Can you explain please?

JoshVanL replied:
Using the [`RunnerManager`](https://github.com/dapr/dapr/blob/master/pkg/concurrency/runner.go#L38) to manage both the `provider` and `server`, ensuring that both return and that both errors are caught.

ItalyPaleAle replied:
Done

JoshVanL reviewed (changes requested):
Thanks @ItalyPaleAle! Some comments from me

ItalyPaleAle commented:
@JoshVanL i’m on vacation at the moment and will be for another 10 days. Feel free to make changes as you wish and merge this (or merge this and make changes as you wish after).

ItalyPaleAle commented:
@JoshVanL i've addressed most of your feedback.

A few things are left, such as the `SENTRY_LOCAL_IDENTITY` env var, which I'm scared of removing until we know it doesn't break stuff.

Also for `DAPR_SENTRY_TOKEN`, I do really like this. I know it's sub-par on K8s, but it's much better outside of K8s, usually much easier than dealing with files.

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +155,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+// Returns the token for authenticating with Sentry.
+func getToken() (token string, validator sentryv1pb.SignCertificateRequest_TokenValidator) {
+	// Check if we have a token in the DAPR_SENTRY_TOKEN env var (for the JWKS validator)
+	if v := os.Getenv(consts.SentryTokenEnvVar); v != "" {
```
We must use this validator if the environment variable set. It's beside the point of the value has any string.

```suggestion
	if v, ok := os.LookupEnv(consts.SentryTokenEnvVar); ok {
```

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +155,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+	}
+
+	// Check if we have a token file in the DAPR_SENTRY_TOKEN_FILE env var (for the JWKS validator)
+	if path := os.Getenv(consts.SentryTokenFileEnvVar); path != "" {
```
Same here. We _must_ return here if the environment is set.

ItalyPaleAle replied:
Done

JoshVanL replied:
We need to check for ok.

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +155,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+		// Errors are logged but we still return the value even if empty
+		b, err := os.ReadFile(path)
+		if err != nil {
+			log.Warnf("Failed to read token at path '%s': %v", path, err)
```
We _must_ return an error here. Silently ignoring this puts us in a wired state and has will have consequences downstream.

The caller can handle waiting for the file to exist in later PRs.

ItalyPaleAle replied:
Done

JoshVanL replied:
@ItalyPaleAle maybe you are missing a push?

JoshVanL commented on pkg/runtime/security/auth.go:
```diff
@@ -153,8 +155,28 @@ func (a *authenticator) CreateSignedWorkloadCert(id, namespace, trustDomain stri
+// Returns the token for authenticating with Sentry.
+func getToken() (token string, validator sentryv1pb.SignCertificateRequest_TokenValidator) {
+	// Check if we have a token in the DAPR_SENTRY_TOKEN env var (for the JWKS validator)
+	if v := os.Getenv(consts.SentryTokenEnvVar); v != "" {
```
I would like to revisit this as I'm not happy that we are accepting static secrets in this way. To recap why:

- Secrets can't be rotated without restarting the Dapr process. This causes churn and downtime. This _will_ happen. Modern JWTs in a cloud native environment should be scoped to 1h -> 24h. With a non-trvial sized cluster (1000s replicas) this is unacceptable.
- Environment variables is public data for all intents and purposes. For 1. they are typically configured along side _config_ which muddles secret and non-secret data. 2. They are published to `/proc` and equivalents, which means the secrets become discoverable in places like metrics, systemd, container runtimes etc. Environment variable values are kept as secret as the name of the process itself.

Putting secrets in environment variables is a massive anti-pattern and wish the Dapr project would set a better example, especially being a modern application. I also don't buy the argument that this is easier for users- so is using hard-coding passwords but people clearly (rightly!) think this is a bad idea. With the security in Dapr not being in a good place right now, I think it would be best to button down the hatches as it were, and make the best decisions from a security perspective. At a later date, we can evaluate where it is worth diminishing the project's security posture in favor of specific user feedback.

ItalyPaleAle replied:
I don't disagree with you with regards to Kubernetes, but I am also thinking of standalone mode, including Docker Compose, development environments, etc. It's a lot more complex to have the files written to disk.

Metrics should not have access to `/proc`, right?

JoshVanL replied:
> , but I am also thinking of standalone mode, including Docker Compose, development environments, etc

I still don't understand this argument. There are first class, easy to use, solutions for this problem. [Docker Compose](https://docs.docker.com/compose/use-secrets/), [podman](https://www.redhat.com/sysadmin/new-podman-secrets-command), [docker](https://docs.docker.com/engine/swarm/secrets/). All dev environments can write files. 

> Metrics should not have access to /proc, right?

This is how machine host metric exports work. https://github.com/prometheus/node_exporter

ItalyPaleAle replied:
Ok, i'll remove the env var

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -102,18 +107,29 @@ func (s *server) SignCertificate(ctx context.Context, req *sentryv1pb.SignCertif
 func (s *server) signCertificate(ctx context.Context, req *sentryv1pb.SignCertificateRequest) (*sentryv1pb.SignCertificateResponse, error) {
-	trustDomain, err := s.val.Validate(ctx, req)
+	validator := s.defaultValidator
+	if req.TokenValidator != 0 && req.TokenValidator.String() != "" {
```
Use the `unkown` constant, rather than `0` here. What's the reason for the string check?

ItalyPaleAle replied:
Done on the constant

String check is to make sure the value is actually defined. For example if the caller is using a newer proto which has more validators allowed than we support.

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -102,18 +107,29 @@ func (s *server) SignCertificate(ctx context.Context, req *sentryv1pb.SignCertif
+	if req.TokenValidator != 0 && req.TokenValidator.String() != "" {
+		validator = req.TokenValidator
+	}
+	if req.TokenValidator == 0 {
```
Again, re constant.

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -125,23 +141,25 @@ func (s *server) signCertificate(ctx context.Context, req *sentryv1pb.SignCertif
+		dns = fmt.Sprintf("%s.%s.svc.cluster.local", req.Id, req.Namespace)
 	}
 
+	log.Infof("Processing SignCertificate requests for %s (validator: %s)", dns, string(validator))
```
Remove. This will print thousands of times and is not necessary.

ItalyPaleAle replied:
I changed it to debug log. It can be useful to see if the request came in while debugging issues.

JoshVanL commented on pkg/sentry/server/validator/internal/common.go:
```diff
@@ -64,3 +64,17 @@ func namespaceIsRequired(namespace string) error {
 	return nil
 }
+
+// IsControlPlaneComponent returns true if the app ID corresponds to a Dapr control plane component.
+// Note: callers must additionally validate the namespace to ensure it matches the one of the Dapr control plane.
+func IsControlPlaneComponent(appID string) bool {
```
```suggestion
// IsControlPlaneApp returns true if the app ID corresponds to a Dapr control plane component.
// Note: callers must additionally validate the namespace to ensure it matches the one of the Dapr control plane.
func IsControlPlaneApp(id string) bool {
```

`App` is semantically probably the right word to be using here.

ItalyPaleAle replied:
I agree with Component not being good. I'll use "Service" however because with "App" there's ambiguity as we normally talk about "apps" in the context of user code.

JoshVanL reviewed (changes requested):
Thanks @ItalyPaleAle. Some more comments from me- particular points of interest is the discussion around env var secrets, and checking the namespace on incoming requests.

@ItalyPaleAle Could you also add integration tests for this? We will be able to tests requests with valid and non-valid JWTs, based on client keys, the contents of the token, and mismatching of token info and request. We can also tests the behavior of validators enabled based on config.

ItalyPaleAle commented:
@JoshVanL I've applied the first round of comments, and I'll do the rest later today.

> Could you also add integration tests for this? We will be able to tests requests with valid and non-valid JWTs, based on client keys, the contents of the token, and mismatching of token info and request. We can also tests the behavior of validators enabled based on config.

Would you mind if I did this in a separate PR? having PRs against PRs open for long is a PITA since there are "2 upstreams" to track.

JoshVanL commented:
> Would you mind if I did this in a separate PR? having PRs against PRs open for long is a PITA since there are "2 upstreams" to track.

@ItalyPaleAle Ideally we keep it to this PR. Both to help everyone start building the muscle of writing integration tests along with feature development, as well as keeping things together in single PRs. PR isn't going anywhere is OK to have it open for longer.

JoshVanL commented on pkg/runtime/security/consts/consts.go:
```diff
@@ -15,10 +15,15 @@ package consts
-	// Header name for HTTP/gRPC calls to hold the token.
+	// SentryLocalIdentityEnvVar is the environment variable for the local identity sent to Sentry.
+	SentryLocalIdentityEnvVar = "SENTRY_LOCAL_IDENTITY"
+	// SentryTokenEnvVar is the environment variable for the Sentry token file.
```
Update comment

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/sentry/server/validator/jwks/jwks.go:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	// Construct the expected value for the subject, which is the SPIFFE ID of the requestor
+	sub, err := spiffeid.FromPathf(td, "/ns/%s/%s", req.Namespace, req.Id)
```
Use `FromSegments` to ensure that neither the namespace or ID contain any separators. 

```suggestion
	sub, err := spiffeid.FromSegments(td, "ns", req.Namespace, req.Id)
```

ItalyPaleAle replied:
Thanks!

ItalyPaleAle commented:
@JoshVanL should be done now. ITs are present too for the insecure and jwks validators (for the k8s validator we'll need to rely on E2E tests when everything is done)

JoshVanL commented on tests/integration/framework/process/sentry/options.go:
```diff
@@ -47,3 +48,9 @@ func WithCABundle(bundle ca.CABundle) Option {
 	}
 }
+
+func WithConfiguration(config *configv1alpha1.ConfigurationSpec) Option {
+	return func(o *options) {
+		o.configuration = config
+	}
+}
```
We should always pass raw test strings for configuration files in integration tests. This ensures that we have proper test coverage and don't miss issues which arise from Marshal. See https://github.com/dapr/dapr/blob/8d7eaeea06541cdec356765dea1911bd95107b18/tests/integration/framework/process/daprd/options.go#L126

ItalyPaleAle replied:
Done

JoshVanL commented on tests/integration/suite/sentry/utils.go:
```diff
@@ -0,0 +1,80 @@
+
+// Keys used to sign and verify JWTs
+const (
+	jwtSigningKeyPubJSON  = `{"kid":"mykey","kty":"EC","alg":"ES256","crv":"P-256","x":"UMn1c2ioMNi2DqvC8hdBVUERFZ97eVFsNVcQIgR0Hso","y":"uT1a0P3UOLiObve2-pOMFx2BVzLz5rFtU-qmQBPWwd0"}`
+	jwtSigningKeyPrivJSON = `{"kid":"mykey","kty":"EC","alg":"ES256","crv":"P-256","d":"5wV7hDpqt1L3uaXa1Xj7X3ieaV9A-Hyj2Kv-qxpwSjM","x":"UMn1c2ioMNi2DqvC8hdBVUERFZ97eVFsNVcQIgR0Hso","y":"uT1a0P3UOLiObve2-pOMFx2BVzLz5rFtU-qmQBPWwd0"}`
```
There is no need for these to be hardcoded. As a rule of thumb, never commit private key material. Please remove and generate the key at test runtime.

ItalyPaleAle replied:
The reason why I hardcoded these is because generating a private key can take time and the ITs need to run quickly.

Although I guess that generating a P-256 key should be relatively fast unlike RSA keys. I'll make the change and hope it doesn't cause issues :)

JoshVanL commented on tests/integration/suite/sentry/utils.go:
```diff
@@ -0,0 +1,80 @@
+var jwtSigningKeyPriv jwk.Key
+
+func init() {
+	jwtSigningKeyPriv, _ = jwk.ParseKey([]byte(jwtSigningKeyPrivJSON))
```
Check errors.

ItalyPaleAle replied:
This cannot fail in this scenario (even if we generate our own key)

JoshVanL commented on tests/integration/suite/sentry/sentry.go:
```diff
@@ -0,0 +1,406 @@
+}
+
+// insecureValidator tests Sentry with the insecure validator.
+type insecureValidator struct {
```
This struct name doesn't make much sense. Remember that the struct name is going to be used as part of the test name also.

ItalyPaleAle replied:
good catch, i should split the test into two structs

JoshVanL commented on tests/integration/suite/sentry/sentry.go:
```diff
@@ -0,0 +1,406 @@
+func (m *insecureValidator) Setup(t *testing.T) []framework.Option {
+	m.sentryWithInsecure = procsentry.New(t)
+
+	jwksValidatorConfig, _ := json.Marshal(map[string]string{
+		"source":             `{"keys":[` + jwtSigningKeyPubJSON + `]}`,
+		"minRefreshInterval": "2m",
+		"requestTimeout":     "1m",
+	})
+	m.sentryWithJWKS = procsentry.New(t, procsentry.WithConfiguration(&configurationv1alpha1.ConfigurationSpec{
+		MTLSSpec: configurationv1alpha1.MTLSSpec{
+			Enabled: true,
+			TokenValidators: []configurationv1alpha1.ValidatorSpec{
+				{
+					Name: "jwks",
+					Options: &configurationv1alpha1.DynamicValue{
+						JSON: apiextensionsv1.JSON{
+							Raw: jwksValidatorConfig,
+						},
+					},
+				},
+			},
+		},
+	}))
```
As mentioned, we should writing our configuration as strings in integration tests.

ItalyPaleAle replied:
Done

JoshVanL commented on tests/integration/suite/sentry/sentry.go:
```diff
@@ -0,0 +1,406 @@
+				TokenValidator:            sentrypbv1.SignCertificateRequest_INSECURE,
+			})
+			require.NoError(t, err)
+			require.NotEmpty(t, res.WorkloadCertificate)
```
Verify the contents of the certificate is correct, i.e. it has been signed by the same Certificate Authority as sentry is serving with, as well as it containing the expected SPIFFE ID of this app.

ItalyPaleAle replied:
Done

JoshVanL commented on tests/integration/suite/sentry/sentry.go:
```diff
@@ -0,0 +1,406 @@
+			_, err := client.SignCertificate(ctx, &sentrypbv1.SignCertificateRequest{
+				Id:                        defaultAppID,
+				Namespace:                 defaultNamespace,
+				CertificateSigningRequest: []byte("-----BEGIN PRIVATE KEY-----\nMC4CAQAwBQYDK2VwBCIEIAYIsKL0xkTkAXDhUN6eDheqODEOGyFZ04jsgFNCFxZf\n-----END PRIVATE KEY-----"),
```
Please don't do this, again re committing private keys is bad. Some random garbage data here is fine.

ItalyPaleAle replied:
Done

JoshVanL commented on tests/integration/suite/sentry/sentry.go:
```diff
@@ -0,0 +1,406 @@
+				Token:                     string(token),
+			})
+			require.NoError(t, err)
+			require.NotEmpty(t, res.WorkloadCertificate)
```
Again re checking the integrity and contents of the certificate.

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -125,23 +141,25 @@ func (s *server) signCertificate(ctx context.Context, req *sentryv1pb.SignCertif
 	switch {
 	case req.Namespace == security.CurrentNamespace() && req.Id == "dapr-injector":
-		dns = []string{fmt.Sprintf("dapr-sidecar-injector.%s.svc", req.Namespace)}
+		dns = fmt.Sprintf("dapr-sidecar-injector.%s.svc", req.Namespace)
 	case req.Namespace == security.CurrentNamespace() && req.Id == "dapr-operator":
-		dns = []string{fmt.Sprintf("dapr-webhook.%s.svc", req.Namespace)}
+		dns = fmt.Sprintf("dapr-webhook.%s.svc", req.Namespace)
 	default:
-		dns = []string{fmt.Sprintf("%s.%s.svc.cluster.local", req.Id, req.Namespace)}
+		dns = fmt.Sprintf("%s.%s.svc.cluster.local", req.Id, req.Namespace)
 	}
 
+	log.Debugf("Processing SignCertificate requests for %s (validator: %s)", dns, validator.String())
+
 	chain, err := s.ca.SignIdentity(ctx, &ca.SignRequest{
 		PublicKey:          csr.PublicKey,
 		SignatureAlgorithm: csr.SignatureAlgorithm,
 		TrustDomain:        trustDomain.String(),
 		Namespace:          req.Namespace,
 		AppID:              req.Id,
-		DNS:                dns,
+		DNS:                []string{dns},
```
Can you please leave this as it was?

The default switch on line 151 is only there for backwards compat, and will be removed in future meaning that an empty list is going to be passed to `SignIdentity` and so we are going to have to change it back anyway.

ItalyPaleAle replied:
Ok I understand.

I did this primarily for logging reasons, but I can move the log inside the switch.

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -44,17 +44,21 @@ type Options struct {
 type server struct {
-	val validator.Validator
-	ca  ca.Signer
+	vals             map[sentryv1pb.SignCertificateRequest_TokenValidator]validator.Validator
```
```suggestion
	validators             map[sentryv1pb.SignCertificateRequest_TokenValidator]validator.Validator
```

`vals` is a bit of an overloaded word. I would think it is a slice integers or something.

ItalyPaleAle replied:
Done

JoshVanL commented on pkg/sentry/sentry.go:
```diff
@@ -104,60 +111,96 @@ func (s *sentry) Start(ctx context.Context) error {
-
-	log.Info("Validator created, starting sentry server")
+	// Start all background processes
+	runners := []concurrency.Runner{
```
```suggestion
	runners := concurrency.NewRunnerManager(
```

JoshVanL commented on pkg/sentry/sentry.go:
```diff
@@ -104,60 +111,96 @@ func (s *sentry) Start(ctx context.Context) error {
+		},
+	}
+	for name, val := range vals {
+		log.Infof("Starting validator %s", strings.ToLower(name.String()))
```
Pedantically, we are not running any validator at this point.

```suggestion
		log.Infof("Using validator %q", strings.ToLower(name.String()))
```

JoshVanL commented on pkg/sentry/sentry.go:
```diff
@@ -104,60 +111,96 @@ func (s *sentry) Start(ctx context.Context) error {
+	}
+	for name, val := range vals {
+		log.Infof("Starting validator %s", strings.ToLower(name.String()))
+		runners = append(runners, val.Start)
```
```suggestion
		runners.Add(val.Start)
```

JoshVanL commented on pkg/sentry/server/server.go:
```diff
@@ -141,25 +141,27 @@ func (s *server) signCertificate(ctx context.Context, req *sentryv1pb.SignCertif
-		dns = fmt.Sprintf("%s.%s.svc.cluster.local", req.Id, req.Namespace)
+		dns = []string{fmt.Sprintf("%s.%s.svc.cluster.local", req.Id, req.Namespace)}
+		log.Debugf("Processing SignCertificate requests for %s (validator: %s)", dns[0], validator.String())
+		log.Warnf("Processing a SignCertificate request for %s using the compatibility method. This will be unsupported in a future version of Sentry. Please upgrade your Dapr sidecars to a more recent release.", dns[0])
```
There is nothing to really warn about here- all sidecars will continue to receive a certificate with DNS SANs regardless of the version the sidecar is running.

```suggestion

```

JoshVanL commented on tests/integration/suite/sentry/utils.go:
```diff
@@ -78,3 +104,38 @@ func signJWT(builder *jwt.Builder) ([]byte, error) {
+		cert, err := x509.ParseCertificate(block.Bytes)
+		require.NoError(t, err)
+
+		assert.Equal(t, []string{expectDNSName}, cert.DNSNames)
```
Also check the SPIFFE ID is correct for the app in the signed certificate. This is contained in the `cert.URIs` field.

JoshVanL replied:
Please also check that the certificate has the `server auth` and `client auth` key usages.

ItalyPaleAle replied:
Done
